// @truffle/preserve-to-buckets/dist/lib/index.js
"use strict";
/**
 * @module @truffle/preserve-to-buckets
 */ /** */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Recipe = void 0;
const Preserve = __importStar(require("@truffle/preserve"));
const clear_1 = require("./clear");
const connect_1 = require("./connect");
const upload_1 = require("./upload");
// @textile/hub requires a WebSocket API to be available on the global object
if (typeof global !== "undefined")
    global.WebSocket = require('isomorphic-ws');
class Recipe {
    constructor(options) {
        this.name = "@truffle/preserve-to-buckets";
        this.inputLabels = ["fs-target"];
        this.outputLabels = ["ipfs-cid"];
        this.key = options.key;
        this.secret = options.secret;
        this.bucketName = options.bucketName;
    }
    async *execute(options) {
        const { inputs, controls } = options;
        const { update } = controls;
        const { key, secret, bucketName } = this;
        const target = Preserve.Targets.normalize(inputs["fs-target"]);
        if (Preserve.Targets.Sources.isContent(target.source)) {
            throw new Error("@truffle/preserve-to-buckets only supports preserving directories.");
        }
        yield* update({ message: "Preserving to Textile Buckets..." });
        const { buckets, bucketKey } = yield* connect_1.connect({ key, secret, bucketName, controls });
        yield* clear_1.clear({ buckets, bucketKey, controls });
        const { cid } = yield* upload_1.upload({ target, buckets, bucketKey, controls });
        return { "ipfs-cid": cid };
    }
}
exports.Recipe = Recipe;
Recipe.help = "Preserve to Textile Buckets";
