// @truffle/preserve-to-filecoin/dist/lib/index.js
"use strict";
/**
 * @module @truffle/preserve-to-filecoin
 */ /** */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Recipe = exports.defaultStorageDealOptions = exports.defaultAddress = void 0;
const Preserve = __importStar(require("@truffle/preserve"));
const connect_1 = require("./connect");
const miners_1 = require("./miners");
const storage_1 = require("./storage");
const wait_1 = require("./wait");
exports.defaultAddress = "http://localhost:7777/rpc/v0";
exports.defaultStorageDealOptions = {
    epochPrice: "250",
    duration: 518400,
};
class Recipe {
    constructor(options) {
        this.name = "@truffle/preserve-to-filecoin";
        this.inputLabels = ["fs-target", "ipfs-cid"];
        this.outputLabels = ["filecoin-deal-cid"];
        this.address = options.address || exports.defaultAddress;
        this.token = options.token;
        this.storageDealOptions = {
            ...exports.defaultStorageDealOptions,
            ...options.storageDealOptions
        };
    }
    async *execute(options) {
        const { address: url, token, storageDealOptions } = this;
        const { inputs, controls } = options;
        const { update } = controls;
        if (Preserve.Targets.Sources.isContent(inputs["fs-target"].source)) {
            throw new Error("@truffle/preserve-to-filecoin only supports preserving directories at this time.");
        }
        const cid = inputs["ipfs-cid"];
        yield* update({ message: "Preserving to Filecoin..." });
        const client = yield* connect_1.connect({ url, token, controls });
        const miners = yield* miners_1.getMiners({ client, controls });
        const { dealCid } = yield* storage_1.proposeStorageDeal({
            cid,
            client,
            storageDealOptions,
            miners,
            controls
        });
        yield* wait_1.wait({ client, dealCid, controls });
        return { "filecoin-deal-cid": dealCid };
    }
}
exports.Recipe = Recipe;
Recipe.help = "Preserve to Filecoin";
